exercises/enumerating_trees/descr.md

```diff
@@ -0,0 +1,364 @@
+# Enumerating Trees
+
+In this exercise, we write code that can **count**, **enumerate**,
+and **sample** the elements of a certain data type.
+This allows us to do things such as:
+
+* Count the binary trees of size `s`.
+* Iterate over all binary trees of size `s`.
+* Pick a binary tree of size `s` uniformly at random among all such trees.
+
+We will also be interested in imposing additional invariants: for instance, we
+can easily restrict our attention to **balanced** binary trees.
+
+This exercise is inspired by the paper
+[Feat: functional enumeration of algebraic types](https://www.semanticscholar.org/paper/Feat%3A-functional-enumeration-of-algebraic-types-Dureg%C3%A5rd-Jansson/46aabcca02c47095b29f934e8d7198bf4c58c27d)
+by Jonas Duregård, Patrik Jansson and Meng Wang.
+
+## Enumerations of Sized Data
+
+Instead of restricting our attention to the type of trees (as in the above
+examples), we would like to work with an arbitrary type `'a` whose values are
+equipped with a notion of **size**. The type `'a` may have an infinite number
+of values; yet, we assume that size is defined in such a way that, for every
+natural number `s`, there is a finite subset of values of type `'a` whose size
+is `s`. We would like to be able to count, enumerate, and sample this subset.
+
+With this in mind, we define an **enumeration** of the type `'a`
+to be a function which, given a *nonnegative* size `s`,
+returns a finite sequence of values of type `'a` whose size is `s`.
+
+```
+  type 'a enum =
+    int -> 'a Seq.seq
+```
+
+Thus, an enumeration `e` of type `'a enum` represents a set of elements of
+type `'a`, grouped by size. The function call `e(s)` returns the subset of
+the elements of size `s`.
```

Comment:
At some point, it could be a good idea to say that the sequence has no repetition.

Reply:
Ok.